multiview/cluster/mv_spectral.py

```diff
@@ -0,0 +1,232 @@
+# limitations under the License.
+#
+# Implements multi-view spectral clustering algorithm for data with
+# multiple views.
+
+
+import numpy as np
+import scipy as sp
+from scipy.spatial.distance import cdist
+from sklearn.base import BaseEstimator
+from sklearn.cluster import KMeans
+from ..utils.utils import check_Xs
+from sklearn.exceptions import NotFittedError
+
+
+class MultiviewSpectralClustering(BaseEstimator):
+
+    '''
+    An implementation of Multi-View Spectral using the
+    basic co-training framework.
+
+    Paramters
+    ---------
+    n_clusters : int
+        The number of clusters
+
+    n_views : int, optional (default=2)
+        The number of different views of data.
+
+    random_state : int, optional (default=None)
+        Determines random number generation for kmeans.
+
+    info_view : int, optional (default=None)
+        The most informative view. Must be between 0 and n_views-1
+        If given, then the final clustering will be performed on the
+        designated view alone. Otherwise, the algorithm will concatenate
+        across all views and cluster on the result.
+
+    max_iter : int, optional (default=10)
+        The maximum number of iterations to run the clustering
+        algorithm.
+
+    References
+    ----------
+    [1] Abhishek Kumar and Hal Daume. A Co-training Approach for Multiview
+    Spectral Clustering. In International Conference on Machine Learning, 2011
+
+    '''
+
+    def __init__(self, n_clusters, n_views=2, random_state=None,
+                 info_view=None, n_iter=10):
+
+        super().__init__()
+
+        if not (isinstance(n_clusters, int) and n_clusters > 0):
+            msg = 'n_clusters must be a positive integer'
+            raise ValueError(msg)
+
+        if not (isinstance(n_views, int) and n_views > 0):
+            msg = 'n_views must be a positive integer'
+            raise ValueError(msg)
+
+        if random_state is not None:
+            msg = 'random_state must be convertible to 32 bit unsigned integer'
+            try:
+                random_state = int(random_state)
+            except ValueError:
+                raise ValueError(msg)
+            np.random.seed(random_state)
+
+        self.info_view = None
+        if info_view is not None:
+            if (isinstance(info_view, int)
+                    and (info_view >= 0 and info_view < n_views)):
+                self.info_view = info_view
+            else:
+                msg = 'info_view must be an integer between 0 and n_clusters-1'
+                raise ValueError(msg)
+
+        if not (isinstance(n_iter, int) and (n_iter > 0)):
+            msg = 'max_iter must be a positive integer'
+            raise ValueError(msg)
+
+        self.n_clusters = n_clusters
+        self.n_views = n_views
+        self.random_state = random_state
+        self.info_view = info_view
+        self.max_iter = n_iter
+
+    def _gaussian_sim(self, X):
```

Comment:
just a note - i realize this is what was done in the paper - but gaussian similarity kernel is fairly restrictive, there are many other kernels one can use. Also, this is already implemented in sklearn, so I don't see why you need to implement it here (more to test, higher chance of bugs). 

I won't ask for it for the final for this semester, but I think addressing both of these comments would be worthwhile in the future. Basically the goal would be make yours have more of the options that sklearn's spectral clustering has

Reply:
also should be able to pick the bandwidth here